File: load-agent/clients/loadgen_client.py
"""Клиент loadgen для Python >= 3.8. Без зависимостей (stdlib).

Тонкая обёртка над load-agent/bin/loadgen.mjs: движок нагрузки один и оттестирован,
клиент только запускает его как процесс и разбирает JSON-результат.
Требуется Node.js >= 18 в PATH (или передайте node="путь/к/node").

Пример (pytest):

    from loadgen_client import run, LoadgenError

    def test_pools_endpoint_holds_load():
        result = run("scenarios/pools.json", duration=30)
        assert result["verdict"] == "PASS", result["hints"]
        assert result["latencyMs"]["p95"] < 200

Поля результата — см. README движка: verdict, latencyMs{p50..max}, rps, errorRatePct,
perRequest[], checksSummary[], paramImpact[], hints[], errorsDetail[].
"""
from __future__ import annotations

import json
import os
import subprocess
import tempfile
import uuid

__all__ = ["run", "probe", "validate", "LoadgenError", "TargetUnreachableError"]

_DEFAULT_LOADGEN = os.path.normpath(
    os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "bin", "loadgen.mjs")
)


class LoadgenError(RuntimeError):
    """Ошибка конфигурации сценария или использования CLI (exit code 1)."""


class TargetUnreachableError(LoadgenError):
    """Цель недоступна или pre-flight (логин/vars) не прошёл (exit code 3)."""


def _exec(args, timeout):
    return subprocess.run(args, capture_output=True, text=True, encoding="utf-8", timeout=timeout)


def probe(base_url, paths=(), *, node="node", loadgen=None, timeout=60):
    """Проверка доступности цели. Возвращает dict {reachable: bool, output: str}."""
    proc = _exec([node, loadgen or _DEFAULT_LOADGEN, "probe", base_url, *paths], timeout)
    return {"reachable": proc.returncode == 0, "output": proc.stdout + proc.stderr}


def validate(scenario, *, node="node", loadgen=None, timeout=60):
    """Валидация сценария. Возвращает dict {ok: bool, output: str}."""
    proc = _exec([node, loadgen or _DEFAULT_LOADGEN, "validate", str(scenario)], timeout)
    return {"ok": proc.returncode == 0, "output": proc.stdout + proc.stderr}


def run(
    scenario,
    *,
    smoke=False,
    vus=None,
    duration=None,
    base_url=None,
    allow_writes=False,
    confirm_external=False,
    out=None,
    node="node",
    loadgen=None,
    timeout=1200,
):
    """Прогон нагрузки. Возвращает dict с полным JSON-результатом движка.

    Дополнительно кладёт в результат exitCode (0 = PASS, 2 = FAIL) и consoleOutput.
    Бросает LoadgenError (кривой сценарий) или TargetUnreachableError (цель лежит).
    Вердикт по порогам НЕ бросает исключение — проверяйте result["verdict"] сами.
    """
    out_path = out or os.path.join(tempfile.gettempdir(), f"loadgen-{uuid.uuid4().hex}.json")
    cmd = [node, loadgen or _DEFAULT_LOADGEN, "run", str(scenario), "--quiet", "--out", out_path]
    if smoke:
        cmd.append("--smoke")
    if vus is not None:
        cmd += ["--vus", str(vus)]
    if duration is not None:
        cmd += ["--duration", str(duration)]
    if base_url is not None:
        cmd += ["--base-url", base_url]
    if allow_writes:
        cmd.append("--allow-writes")
    if confirm_external:
        cmd.append("--confirm-external")

    proc = _exec(cmd, timeout)
    console = proc.stdout + proc.stderr
    if proc.returncode == 1:
        raise LoadgenError(console.strip())
    if proc.returncode == 3:
        raise TargetUnreachableError(console.strip())

    try:
        with open(out_path, encoding="utf-8") as f:
            result = json.load(f)
    except FileNotFoundError as e:
        raise LoadgenError(f"Движок не записал результат ({out_path}). Вывод:\n{console}") from e
    finally:
        if out is None and os.path.exists(out_path):
            os.unlink(out_path)

    result["exitCode"] = proc.returncode
    result["consoleOutput"] = console
    return result

File: load-agent/clients/loadgen-client.mjs
/**
 * Клиент loadgen для Node.js >= 18 (программное использование из тестов/скриптов).
 * Без зависимостей. Движок — ../bin/loadgen.mjs, клиент только запускает процесс.
 *
 * Пример (node:test):
 *   import { run } from '../clients/loadgen-client.mjs';
 *   const r = await run('scenarios/pools.json', { duration: 30 });
 *   assert.equal(r.verdict, 'PASS');
 */
import { spawn } from 'node:child_process';
import { readFileSync, unlinkSync, existsSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { tmpdir } from 'node:os';
import { randomUUID } from 'node:crypto';

const DEFAULT_LOADGEN = join(dirname(fileURLToPath(import.meta.url)), '..', 'bin', 'loadgen.mjs');

export class LoadgenError extends Error {
  constructor(message, exitCode) { super(message); this.exitCode = exitCode; }
}

function exec(args, timeoutMs) {
  return new Promise((resolve, reject) => {
    const p = spawn(process.execPath, args, { stdio: ['ignore', 'pipe', 'pipe'] });
    let out = '';
    p.stdout.on('data', (d) => { out += d; });
    p.stderr.on('data', (d) => { out += d; });
    const t = setTimeout(() => { p.kill(); reject(new LoadgenError(`loadgen не завершился за ${timeoutMs}мс`, -1)); }, timeoutMs);
    p.on('close', (code) => { clearTimeout(t); resolve({ code, out }); });
    p.on('error', (e) => { clearTimeout(t); reject(e); });
  });
}

/** Проверка доступности цели: { reachable, output } */
export async function probe(baseUrl, paths = [], { loadgen = DEFAULT_LOADGEN, timeoutMs = 60_000 } = {}) {
  const { code, out } = await exec([loadgen, 'probe', baseUrl, ...paths], timeoutMs);
  return { reachable: code === 0, output: out };
}

/** Валидация сценария: { ok, output } */
export async function validate(scenario, { loadgen = DEFAULT_LOADGEN, timeoutMs = 60_000 } = {}) {
  const { code, out } = await exec([loadgen, 'validate', String(scenario)], timeoutMs);
  return { ok: code === 0, output: out };
}

/**
 * Прогон нагрузки. Возвращает полный JSON-результат движка + exitCode + consoleOutput.
 * Бросает LoadgenError при кривом сценарии (exit 1) и недоступной цели (exit 3);
 * вердикт FAIL по порогам исключение НЕ бросает — проверяйте result.verdict.
 */
export async function run(scenario, opts = {}) {
  const {
    smoke = false, vus, duration, baseUrl,
    allowWrites = false, confirmExternal = false,
    out, loadgen = DEFAULT_LOADGEN, timeoutMs = 1_200_000,
  } = opts;
  const outPath = out || join(tmpdir(), `loadgen-${randomUUID()}.json`);
  const args = [loadgen, 'run', String(scenario), '--quiet', '--out', outPath];
  if (smoke) args.push('--smoke');
  if (vus !== undefined) args.push('--vus', String(vus));
  if (duration !== undefined) args.push('--duration', String(duration));
  if (baseUrl !== undefined) args.push('--base-url', baseUrl);
  if (allowWrites) args.push('--allow-writes');
  if (confirmExternal) args.push('--confirm-external');

  const { code, out: console_ } = await exec(args, timeoutMs);
  if (code === 1 || code === 3) throw new LoadgenError(console_.trim(), code);
  try {
    const result = JSON.parse(readFileSync(outPath, 'utf8'));
    result.exitCode = code;
    result.consoleOutput = console_;
    return result;
  } finally {
    if (!out && existsSync(outPath)) unlinkSync(outPath);
  }
}
